### PR TITLE
upgrade: New API to set/get the current upgrade mode

### PR DESCRIPTION
**Pull request by rhafer**
Base on the recent status library changes it is now possible to request
a certain upgrade mode using POST/GET on /api/upgrade/mode.

### CODE REVIEW

File: crowbar_framework/app/controllers/api/upgrade_controller.rb [hunk 1]
```diff
@@ -227,6 +227,28 @@ def openstackbackup
...
+      }
+    end
+  rescue ::Crowbar::Error::SaveUpgradeModeError,
+         ::Crowbar::Error::SaveUpgradeStatusError,
+         ::Crowbar::Error::UpgradeError => e
+    render json: {
+      errors: {
+        mode: {
+          data: e.message,
+          help: "Please refer to the error message in the response."
```
Comment (jsuchome) [inline crowbar_framework/app/controllers/api/upgrade_controller.rb]: I think we should just drop the `help` part if it does not include anything interesting

File: crowbar_framework/app/models/api/upgrade.rb [hunk 2]
```diff
@@ -30,6 +30,17 @@ def node_status
...
 
+      def upgrade_mode
+          ::Crowbar::UpgradeStatus.new.upgrade_mode
+      end
+
+      def upgrade_mode=(mode)
+        unless ["normal", "non_disruptive"].include? mode
+          Rails.logger.debug("Setting upgrade mode #{mode}")
+          ::Crowbar::UpgradeStatus.new.save_selected_upgrade_mode(mode.to_sym)
+        end
```
Comment (jsuchome) [inline crowbar_framework/app/models/api/upgrade.rb]: What if user passes different value? There should be some help message showing allowed values in such case...

File: crowbar_framework/app/controllers/api/upgrade_controller.rb (hunk 1 above)
Comment (skazi0) [inline crowbar_framework/app/controllers/api/upgrade_controller.rb]: +1

Comment (rhafer) [inline crowbar_framework/app/controllers/api/upgrade_controller.rb]: done

File: crowbar_framework/app/models/api/upgrade.rb (hunk 2 above)
Comment (rhafer) [inline crowbar_framework/app/models/api/upgrade.rb]: True. Added that. And also a unit test for it.

File: crowbar_framework/app/models/api/upgrade.rb [hunk 3]
```diff
@@ -30,6 +30,18 @@ def node_status
         }
       end
 
+      def upgrade_mode
+        ::Crowbar::UpgradeStatus.new.upgrade_mode
+      end
+
+      def upgrade_mode=(mode)
+        unless ["normal", "non_disruptive"].include? mode
+          raise ::Crowbar::Error::SaveUpgradeModeError, "Invalid upgrade mode #{mode}."
```
Comment (jsuchome) [inline crowbar_framework/app/models/api/upgrade.rb]: Could we please print allowed values here?

Comment (rhafer) [inline crowbar_framework/app/models/api/upgrade.rb]: Hm, I actually added that to crowbar-client already (the PR for that is will follow soon). Buy yeah, I can add it here as well.

Comment (rhafer) [inline crowbar_framework/app/models/api/upgrade.rb]: Done

**Comment by rhafer** [comment]
[(3) S56P4 API for forcing disruptive upgrade mode](https://trello.com/c/XY1ZBnMg/319-3-s56p4-api-for-forcing-disruptive-upgrade-mode)

**Pull request merged**